helpers.js

```diff
@@ -1,65 +1,194 @@
+const _ = require("lodash");
+const fs = require("fs");
+const { removeUndefinedAndEmpty } = require("kaholo-aws-plugin/helpers");
+
 const GRANTEE_TYPE_TO_FIELD = {
-    "CanonicalUser": "ID",
-    "AmazonCustomerByEmail": "EmailAddress",
-    "Group": "URI"
+  CanonicalUser: "ID",
+  AmazonCustomerByEmail: "EmailAddress",
+  EmailAddress: "EmailAddress",
+  Group: "URI",
+};
+
+function getGranteeId(grant) {
+  return grant.Grantee[GRANTEE_TYPE_TO_FIELD[grant.Grantee.Type]];
+}
+
+function parseGrantee(grantee, granteeType) {
+  const valField = GRANTEE_TYPE_TO_FIELD[granteeType];
+  return {
+    Type: granteeType,
+    [valField]: grantee,
+  };
 }
 
-function parseGrantees(grantees, granteeType){
-    return grantees.map((grantee) => parseGrantee(grantee, granteeType));
+function parseGrantees(grantees, granteeType) {
+  return grantees.map((grantee) => parseGrantee(grantee, granteeType));
 }
 
-function parseGrantee(grantee, granteeType){
-    const valField = GRANTEE_TYPE_TO_FIELD[granteeType];
-    granteeObj = {
-        Type: granteeType
-    }
-    granteeObj[valField] = grantee;
-    return granteeObj;
+function getGrants(grantees, permissionTypes) {
+  return _.flatten(permissionTypes.map((permissionType) => grantees.map((grantee) => ({
+    Grantee: grantee,
+    Permission: permissionType,
+  }))));
 }
 
-function getAwsCallback(resolve, reject){
-    return (err, data) => {
-        if (err) return reject(err);
-        return resolve(data);
-    }
+function combineGrants(currentGrants, newGrants) {
+  if (currentGrants.length === 0) { return newGrants; }
+
+  const grantsAreEqual = (a, b) => a.Permission === b.Permission
+        && a.Grantee.Type === b.Grantee.Type
+        && getGranteeId(a) === getGranteeId(b);
+
+  const applicableNewGrants = newGrants.reduce(
```

Comment:
Why not just `newGrants.filter(newGrant => currentGrants.every(currentGrant => !grantsAreEqual(newGrant, currentGrant))`
or an alternative version depending which case is more likely
`newGrants.filter(newGrant => !currentGrants.some(currentGrant => grantsAreEqual(newGrant, currentGrant))`

Reply:
I 100% agree, I overengineered this one.